### ISSUE TITLE
Add support for multiple deployments in parallel in AzureRmWebAppDeploymentV4

### ISSUE THREAD

**Issue by MauriRojas** [microsoft/azure-pipelines-tasks#10910]
## Required Information

**Question, Bug, or Feature?**  
*Type*: Feature

**Enter Task Name**: AzureRmWebAppDeployment

## Environment
- Server - Azure Pipelines or TFS on-premises?
    
    - Azure Pipelines

- Agent - Hosted or Private: 

    - Private with Windows

## Issue Description

Currently, the AzureRmWebAppDeploymentV4 task allows to deploy to a single Azure App Service. We created a pull request to update it to allow parallel deployments to multiple Azure App Services. Link to the PR: [10909](https://github.com/microsoft/azure-pipelines-tasks/pull/10909)
This pull request adds support for deploying to multiple Azure App Services in one task execution. The services are specified in the App Service name field of the task separated by commas.
![image](https://user-images.githubusercontent.com/31770211/61311591-05443080-a7cd-11e9-92ab-3e16625b2aeb.png)

The task reads the list of names and deploys the app to the services in parallel.

A pre-release version was tested locally and in Azure Pipelines. The names of the apps were hidden from the picture:
![image](https://user-images.githubusercontent.com/31770211/61311489-d6c65580-a7cc-11e9-94af-423fdfca9f2c.png)

Thank you for your time.

**Comment by ArieHein** [comment]
hmm I might be missing something but what prevents you from using the multi-configuration matrix, using a variable to list all the webapps names, and then have one agent job that uses the variable for a web app deploy. This means instead of creating 4 stages, each for a separate web app which is basically a duplication (which you can also solve as a task group, but that's a different topic), you use the existing structure to do the same stage X amount of times based on how many items are in the variable. This is a MUCH better approach then specifically fixing the task as it means additional tasks that you would run in that same stage, can follow the same logic.

In the Agent job section, scroll down to 'Execution Plan', under 'Paralelism' choose 'Multi-Configuration', under 'Multiplyers' , type WebAppName and create a variable with that name that has a comma-separated list of values WebApp1,WebApp2,WebApp3,WebApp4. Choose 'Max number of Agents' as 4 so all 4 values will be run in parallel.

Add powershell task with inline code - Write-Host $(WebAppName)

Que the release.

You will see that 4 jobs will appear, each for a different value in the WebAppName and in each PowerShell task you will see it emitting the appropriate name of the web app as well.


for easy visualization I created this in my pipeline. Enjoy !

![agent_multi_config](https://user-images.githubusercontent.com/15635882/61400360-61dc4400-a8d0-11e9-9bf5-38ff7a13d786.jpg)
![ps_multi_config](https://user-images.githubusercontent.com/15635882/61400361-6274da80-a8d0-11e9-959c-8e0ec9744a96.jpg)
![release_multi_config](https://user-images.githubusercontent.com/15635882/61400362-6274da80-a8d0-11e9-822e-ecb1d6ca5b7d.jpg)
![variable_multi_config](https://user-images.githubusercontent.com/15635882/61400363-6274da80-a8d0-11e9-8cc0-159b8dee3c8a.jpg)



https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-configuration

**Comment by MauriRojas** [comment]
Thank you for your response @ArieHein. We tested this approach already but unfortunately it didn't work for us. One of the reasons is that we have 15 regions to deploy, so during execution there would be 15 agents taken that could be used for another work (we have +10 projects in our organization all with CI configurations for PRs and branches).
About this: `This is a MUCH better approach then specifically fixing the task as it means additional tasks that you would run in that same stage, can follow the same logic.` I think I'm not following, we would use only one task that deploys all 15 regions in parallel in one agent.

Sorry for the late reply and thank you for your time.

**Comment by ArieHein** [comment]
The number of agents it will use is controlled by the max number of agents. It is true that if you want ALL to change at "same time", you will need 15 agents, but is a parallel agent really a cost factor ?

https://marketplace.visualstudio.com/items?itemName=ms.build-release-hosted-pipelines&ssr=false#pricing

deploying a webapp, should be about 5 min, so 15 agents busy for 5 minutes, with the first ones to finish, releasing their agent to other usages..and you cant allow projects to wait for those 5 minutes, enough not to buy 10 more agents ? 

If I understand you correctly, you have 15 webapps, each in a different region you want to update at the same time. Other then making your web app down for the period of time it takes all of them to finish, are your 15 webapps in one resource group or each in a different one? do you use slots or do you accept the downtime ?  I assume they all share the same data layer, what happens when one or more webapps finishes updating before the others in terms of the data ? what if one of them fails for some reason ? do all webapps share a application gateway or load balancer ?

My comment about additional tasks is that the webapp deploy isnt the only task you should have.
Normaly you will have additional tasks like changing variables in config files, versioning, connection strings, adding a tag to the release branch, updating a workitem, etc. When you stick to one task for one resource its easy to add additional tasks that complement or augment parts of the artifact prior or post deployment of the web app, or add tasks that are run if the previous task failed for rollback (for example when you have slots, an error in deploying reverts to the older slot).

Notice I said "same time" at the top, even if youre using the await and async, the execution will not be parallel but im more afraid of a case where one of the webapp deployment will fail or multiple of them while other succedd. I would more go for canary approach of 2 web apps in, then 4 more then 9 last or some combination based on georgraphy which means you dont use all agents at same time and still have time to revert back.

I have roughly 200 projects with about 400 build defintiions and 200 release definitions, where we deploy to azure. Were sitting at 25 parallel jobs and starting to see the limits of it so scaling to 50 soon. we have applications with mutiple global wide web app deployments, and we mostly use slots to allow us to onboard the new version perior to the actuall switch of the slots to new version which means we decoupled the loading the new ver to the actual time we want to do the change while still maintaining low downtime. For lower environments where we dont want or need the slots, using the multi-configuration and limiting the max agents is perfect as the downtime on lower env is nelegent for us.

Add usage of the conditions in a task, you can gain another degree of control where one release triggers anotehr release automatically or other conditions to decide when to run a task or not.

**Comment by MauriRojas** [comment]
Thank you for your answer @ArieHein. We decided we are implementing a different approach where we create multiple agents in a single VM that we will only use for certain deployments. This will solve the problem of having to stop builds/releases when all agents are taken and we need to make a high priority deployment.
I'm closing this since is no longer needed.
